theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,209 @@
+        for( int j = 0; j < cols; ++j )
+        {
+            int label = *( (int *) PyArray_GETPTR2( label_matrix, i, j ) );
+            if ( label >= 0 )  // negative values are assumed to be padding
+            {
+                (*flat_labels)[ label_index++ ] = label;
+                ++label_length;
+            }
+        }
+        (*label_lengths)[ i ] = label_length;
+    }
+}
+
+#section support_code_apply
+
+int APPLY_SPECIFIC(ctc_cost_cpu)(PyArrayObject *  in_activations,
+                                 PyArrayObject *  in_labels,
+                                 PyArrayObject *  in_input_lengths,
+                                 PyArrayObject ** out_costs,
+                                 PyArrayObject ** out_gradients)
+{
+    // setup CTC computation parameters
+    ctcOptions ctc_options;
+    memset( &ctc_options, 0, sizeof(ctcOptions) );
+    ctc_options.loc = CTC_CPU;
+    ctc_options.num_threads = 1;
+
+    npy_float32 * activations = NULL;
+    PyArrayObject * activations_copy = NULL;
+
+    if ( PyArray_IS_C_CONTIGUOUS( in_activations ) )
+    {
+        activations = (npy_float32 *) PyArray_DATA( in_activations );
+    }
+    else
+    {
+        activations_copy = PyArray_GETCONTIGUOUS( in_activations );
+        if ( NULL != activations_copy )
+        {
+            activations = (npy_float32 *) PyArray_DATA( activations_copy );
+        }
+        else
+        {
+            PyErr_Format( PyExc_ValueError,
+                          "Could not create a contiguous copy of activations array." );
+            return 1;            
+        }
+    }
+
+    int * input_lengths = NULL,
+        * flat_labels = NULL,
+        * label_lengths = NULL;
+
+    create_contiguous_input_lengths( in_input_lengths, &input_lengths );
+
+    if ( NULL == input_lengths )
+    {
+        PyErr_Format( PyExc_ValueError,
+                      "Could not allocate storage for input lengths" );
+        return 1;        
+    }
+
+    // flatten labels to conform with library memory layout
+    create_flat_labels( in_labels, &flat_labels, &label_lengths );
+
+    if ( ( NULL == label_lengths ) || ( NULL == flat_labels ) )
+    {
+        PyErr_Format( PyExc_ValueError,
+                      "Could not allocate storage for labels and their lengths" );
+        return 1;
+    }
+
+    npy_int minibatch_size = PyArray_DIMS( in_activations )[1];
+    npy_int alphabet_size = PyArray_DIMS( in_activations )[2];
+
+    void * ctc_cpu_workspace = NULL;
+
+    npy_float32 * costs = NULL;
+    npy_intp cost_size = minibatch_size;
+
+    if ( NULL == (*out_costs) )
+    {
+        // Symbolic variable has no memory backing, so we create one
+        *out_costs = (PyArrayObject *) PyArray_ZEROS( 1, &cost_size, NPY_FLOAT32, 0 );
+    }
+    else if ( PyArray_NDIM( *out_costs ) != 1 ||
+              PyArray_DIMS( *out_costs )[0] != cost_size )  // matrix has the wrong size
+    {
+        Py_XDECREF( *out_costs ); 
```

Comment:
Note that `Py_XDECREF(pointer)` works also with NULL `pointer` (it does nothing in such case). So, this 
`if(A) {allocate} else if(B) {xdecref then allocate}` 
may be compacted to 
`if (A or B) {xdecref then allocate}`.